src/Npgsql/Internal/TypeInfoMapping.cs

```diff
@@ -273,10 +273,11 @@ static TypeInfoFactory CreateComposedFactory(Type mappingType, TypeInfoMapping i
 
     Func<Type?, bool> GetArrayTypeMatchPredicate(Func<Type?, bool> elementTypeMatchPredicate)
         => type => type is null ? elementTypeMatchPredicate(null) : type.IsArray && elementTypeMatchPredicate.Invoke(type.GetElementType()!);
-    Func<Type?, bool> GetListTypeMatchPredicate(Func<Type?, bool> elementTypeMatchPredicate)
-        => type => type is null ? elementTypeMatchPredicate(null) : type.IsConstructedGenericType && type.GetGenericTypeDefinition() is { } def
-            && (def == typeof(List<>) || def == typeof(IList<>))
-            && elementTypeMatchPredicate(type.GetGenericArguments()[0]);
+    Func<Type?, bool> GetListTypeMatchPredicate<TElement>(Func<Type?, bool> elementTypeMatchPredicate)
+        => type => type is null ? elementTypeMatchPredicate(null)
+            // We anti-constrain on IsArray to avoid matching byte/sbyte, short/ushort int/uint
+            // with the list mapping of the earlier type when an exact match is probably available.
+            : !type.IsArray && typeof(IList<TElement>).IsAssignableFrom(type) && elementTypeMatchPredicate(typeof(TElement));
```

Comment:
nit: can also slightly clean up/align GetArrayTypeMatchPredicate just above if you want (e.g. with TElement, remove Invoke)

Reply:
Done